### PR TITLE
[Core][Geometry] remove inheritance from PointerVector

### PR DESCRIPTION
**Pull request by philbucher**
as discussed with @pooyan-dadvand 

this PR removes the inheritance of `Geometry` from `PointerVector`

This is one of the prerequisites for the NURBS stuffs in Kratos

@pooyan-dadvand this seems to work and the Tests in the Main-Applications are passing. There are NO other changes required, i.e. no changes required in Elements, Condition etc.

(The change in the Quad is only a minor fix and is a separate PR => #4870. It is included here bcs it is necessary to make it compile)

### CODE REVIEW

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -389,6 +385,218 @@ class Geometry : public PointerVector<TPointType>
         return *this;
     }
 
+     operator PointsArrayType&()
```
Comment (philbucher) [inline kratos/geometries/geometry.h:383]: @pooyan-dadvand this is the operator that we discussed this morning in order to make it compatible with the rest of the code

File: kratos/geometries/geometry.h [hunk 2]
```diff
@@ -389,6 +385,218 @@ class Geometry : public PointerVector<TPointType>
...
+    //            (*mpPointVector).end(),
+    //            (*r.mpPointVector).begin(),
+    //            this->EqualKeyTo());
+    //}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    iterator                   begin()
```
Comment (philbucher) [inline kratos/geometries/geometry.h]: @pooyan-dadvand how many of the operations that are copied from the PointerVector should remain here?
according to @tteschemacher many of them are used somewhere but I am not sure if `Geometry` should have e.g. a `ptr_rend` method

Maybe we could deprecate the ones that we decide to drop?

File: kratos/geometries/geometry.h [hunk 3]
```diff
@@ -389,6 +385,218 @@ class Geometry : public PointerVector<TPointType>
...
+    {
+        return (*mpPointVector)(i);
+    }
+
+    ConstPointPointerType& operator()(const SizeType& i) const
+    {
+        return (*mpPointVector)(i);
+    }
+
+    //bool operator==(const Geometry& r) const // nothrow
```
Comment (philbucher) [inline kratos/geometries/geometry.h]: is this needed @pooyan-dadvand ?

File: kratos/geometries/geometry.h [hunk 4]
```diff
@@ -2679,23 +2887,22 @@ class Geometry : public PointerVector<TPointType>
 
     GeometryData const* mpGeometryData;
 
+    Kratos::shared_ptr<PointsArrayType> mpPointVector;
```
Comment (philbucher) [inline kratos/geometries/geometry.h]: @pooyan-dadvand 
@tteschemacher told me that you agreed here to use a shared ptr

**Review by pooyan-dadvand — COMMENTED** [review]
Thanks for fast implementation. The only thing that I think @tteschemacher misunderstood me is about the mpPoints pointer which I would go for the normal one:
* If the geometry is not composites like all of the current geometries: It creates the array with new and delete it in destructor
* If the geometry is composite: It only points to its geometry mpPoint without deleting it (or construct a new one if is necessary and delete it)

About the iterators I would deprecate all the reverse ones

**Comment by tteschemacher** [comment]
I don't fully understand your point @pooyan-dadvand . So you don't want the pointer to mpPointVector but instead a vector of pointers. This would require to make all access/ iterator functions virtual. Composites would then just pass through. However, we would not have to have removed the inheritance, but could have made the base functions virtual in the PointerVector. I don't see the advantage in this procedure.

**Comment by pooyan-dadvand** [comment]
I would go for:

`PointsArrayType* mpPointVector;`

instead of:

`Kratos::shared_ptr<PointsArrayType> mpPointVector;`

But the vector has pointers to points as before

**Comment by philbucher** [comment]
@pooyan-dadvand just for me to understand correctly, you only would replace the shared_ptr with a raw pointer and then do the mem-management ourselves?

**Comment by pooyan-dadvand** [comment]
Yes, we manage the memory management by ourselves for allocating the PointerVector but the Points will be stored with shared pointer as before

**Comment by tteschemacher** [comment]
@pooyan-dadvand I changed to raw pointers. Is this as you expected it?

If yes, we will have to delete all destructors in the geometry subclasses. On the other hand, for the composites we will have to ensure that the desctructor is overriden.

**Comment by pooyan-dadvand** [comment]
@tteschemacher I see your point that all destructors should be overridden. So I would go for changing the mpPointsVector to mPoints without using the pointer and having the pointer only in composite material.

**Comment by tteschemacher** [comment]
@pooyan-dadvand so you say we should copy all stuff from PointerVector except for mData, which we will make protected instead of private? 
This we will override then with the operator, something like this?

```
mData operator TContainerType&()
{
    return *(mpGeometry->mpPoints);
}
```

**Comment by philbucher** [comment]
I think this can be closed now @tteschemacher @pooyan-dadvand ?

**Comment by tteschemacher** [comment]
This is to be merged.

**Comment by pooyan-dadvand** [comment]
What you mean about closing it @philbucher?

I thought we should merge this so @tteschemacher can merge his branch with changes of geometrical data. Isn't it?

**Comment by tteschemacher** [comment]
@RiccardoRossi could you please also review this pull request as it is affecting many things.

Thanks!

ping @pooyan-dadvand

File: scripts/build/appveyor/configure_appveyor_VS2017.bat [hunk 5]
```diff
@@ -11,13 +11,13 @@ cmake .. -G "Visual Studio 15 2017 Win64"
 -DCONTACT_STRUCTURAL_MECHANICS_APPLICATION=ON                                   ^
 -DSWIMMING_DEM_APPLICATION=ON                                                   ^
 -DMESH_MOVING_APPLICATION=ON                                                    ^
--DSOLID_MECHANICS_APPLICATION=ON                                                ^
--DCONSTITUTIVE_MODELS_APPLICATION=ON                                            ^
--DDELAUNAY_MESHING_APPLICATION=ON                                               ^
--DCONTACT_MECHANICS_APPLICATION=ON                                              ^
--DPFEM_APPLICATION=ON                                                           ^
--DPFEM_SOLID_MECHANICS_APPLICATION=ON                                           ^
--DPFEM_FLUID_DYNAMICS_APPLICATION=ON                                            ^
```
Comment (pooyan-dadvand) [inline scripts/build/appveyor/configure_appveyor_VS2017.bat]: Unintented change?

File: kratos/tests/cpp_tests/utilities/test_sensitivity_builder.cpp [hunk 6]
```diff
@@ -38,7 +38,7 @@ class TestAdjoint : public TElement
     typedef Kratos::unique_ptr<TestAdjoint> UniquePointer;
 
 
-    static TestAdjoint::Pointer Create(std::size_t NewId, const PointerVector<Node<3>>& rNodes)
```
Comment (pooyan-dadvand) [inline kratos/tests/cpp_tests/utilities/test_sensitivity_builder.cpp]: As a curiosity, why the typename is needed now?

**Review by pooyan-dadvand — COMMENTED** [review]
In general I would approve it but as I was very involved in this implementation I would like someone else approve it.

**Review by RiccardoRossi — COMMENTED** [review]
fine to me.

@KratosMultiphysics/technical-committee if you all give thumb up i'll approve (or approve yourselves to not wait for me)

File: scripts/build/appveyor/configure_appveyor_VS2017.bat (hunk 5 above)
Comment (tteschemacher) [reply scripts/build/appveyor/configure_appveyor_VS2017.bat]: less than you think. But now I guess I should revert it.

File: kratos/tests/cpp_tests/utilities/test_sensitivity_builder.cpp (hunk 6 above)
Comment (tteschemacher) [reply kratos/tests/cpp_tests/utilities/test_sensitivity_builder.cpp]: It was coming from a error massage in Windows.

**Comment by rubenzorrilla** [comment]
Fine for me. It implements what we decided.

**Review by pooyan-dadvand — DISMISSED** [review]
👍

**Comment by tteschemacher** [comment]
@loumalouomega, do you know what's wrong here? It is failing in your application.

**Comment by loumalouomega** [comment]
> @loumalouomega, do you know what's wrong here? It is failing in your application.

do you mean the compilation?

**Comment by loumalouomega** [comment]
It is something from my side:

"C:\projects\kratos\cmake_build\kratos\all_unity.vcxproj" (default target) (1) ->
"C:\projects\kratos\cmake_build\applications\ContactStructuralMechanicsApplication\KratosContactStructuralMechanicsApplication_unity.vcxproj" (default target) (14) ->
"C:\projects\kratos\cmake_build\applications\ContactStructuralMechanicsApplication\KratosContactStructuralMechanicsCore_unity.vcxproj" (default target) (15) ->
(ClCompile target) -> 
  C:\projects\kratos\applications\ContactStructuralMechanicsApplication\custom_utilities\interface_preprocess.cpp(272): error C2664: 'Kratos::Condition::Pointer Kratos::Condition::Create(Kratos::Condition::IndexType,std::shared_ptr<Kratos::Geometry<Kratos::GeometricalObject::NodeType>>,Kratos::Properties::Pointer) const': cannot convert argument 2 from 'const Kratos::InterfacePreprocessCondition::GeometryType' to 'const Kratos::Condition::NodesArrayType &' [C:\projects\kratos\cmake_build\applications\ContactStructuralMechanicsApplication\KratosContactStructuralMechanicsCore_unity.vcxproj]

**Comment by tteschemacher** [comment]
Yes, because it is failing in your app. I am not sure what's exactly wrong...

**Comment by loumalouomega** [comment]
> Yes, because it is failing in your app. I am not sure what's exactly wrong...

Looks like it doesn' accept a geometry pointer directly

**Comment by tteschemacher** [comment]
Can we add a function for this in the geometry base class?

**Comment by loumalouomega** [comment]
Should work now, just needed to call the Points() method

**Comment by tteschemacher** [comment]
Thanks!! Let's see...

**Comment by loumalouomega** [comment]
> Thanks!! Let's see...

imagen

**Comment by philbucher** [comment]
Hm seems like the create fct that directly accepts a Geometry is broken, I think we can easily fix that

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

File: applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp [hunk 7]
```diff
@@ -269,7 +269,7 @@ void InterfacePreprocessCondition::CreateNewCondition(
 {
     KRATOS_TRY;
 
-    Condition::Pointer p_cond = Condition::Pointer(rCondition.Create(ConditionId, rGeometry, pThisProperties));
+    Condition::Pointer p_cond = Condition::Pointer(rCondition.Create(ConditionId, rGeometry.Points(), pThisProperties));
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp:272]: @tteschemacher I would still try to fix this

**Comment by philbucher** [comment]
@tteschemacher I think this was not fully ready, the `Create` should have been fixed properly

also this broke the HDFapp (FYI @msandre )

File: applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp (hunk 7 above)
Comment (tteschemacher) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp:272]: Shall we do this in a new PR?

**Comment by tteschemacher** [comment]
Do you know why HDFapp breaks?

File: applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp (hunk 7 above)
Comment (pooyan-dadvand) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp:272]: I think the const version of the `pointsArrayType&()` is missing

**Comment by pooyan-dadvand** [comment]
Repeating:

> I think the const version of the pointsArrayType&() is missing

**Comment by jcotela** [comment]
@tteschemacher @philbucher @msandre Regarding the HDF5 test, it breaks because this PR changed `Geometry::ptr_begin()` and `Geometry::ptr_end()`. I'm testing a hotfix and will push it shortly.

File: applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp (hunk 7 above)
Comment (philbucher) [reply applications/ContactStructuralMechanicsApplication/custom_utilities/interface_preprocess.cpp:272]: > Shall we do this in a new PR?

yes

**Pull request merged**